### PR TITLE
Sync form factor specific favorites

### PR DESCRIPTION
**Pull request by ayoy**
Task/Issue URL: https://app.asana.com/0/1201493110486074/1204926049616866/f
Tech Design URL: https://app.asana.com/0/481882893211075/1205418608802079/f

**Description**:
This change introduces form factor specific favorites to be used when Sync is enabled.
1 favorites folder is replaced by 3 folders – mobile, desktop and unified.
Users always see only 1 folder and it's their form-factor-specific one. Only users with Sync enabled
get to choose whether they want to see the form-factor-specific favorites or unified (desktop + mobile)
favorites. The setting itself is synced between devices in the Sync account.
BookmarkEntity's isFavorite is replaced with isFavorite(on:) taking folder as a parameter.
API for adding to favorites and removing from favorites was updated to take multiple favorites
folders as needed.
FavoritesDisplayMode enum is introduced to manage display mode in clients.
Bookmarks and Favorites related view models used on iOS are updated to take favorites display mode
and reload their data as the mode changes.
In SyncDataProviders, an abstract SettingSyncHandler class was added to support adding
an arbitrary setting (key-value pair) to Sync. It's subclassed by EmailProtectionSyncHandler
and FavoritesDisplayModeSyncHandlerBase (that needs to be subclassed in client apps due to differences
in User Defaults storage between iOS and macOS apps).

**Steps to test this PR**:
Refer to BSK PR for testing steps.

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16

**Theme Testing**:

* [ ] Light theme
* [ ] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Core/BookmarksExporter.swift [hunk 1]
```diff
@@ -64,7 +64,7 @@ public struct BookmarksExporter {
                 content.append(Template.bookmark(level: level,
                                                  title: entity.title!.escapedForHTML,
                                                  url: entity.url!,
-                                                 isFavorite: entity.isFavorite))
+                                                 isFavorite: entity.isFavorite(on: .mobile)))
```
Comment (ayoy) [inline Core/BookmarksExporter.swift]: Parametrize this

File: Core/SyncDataProviders.swift [hunk 2]
```diff
@@ -71,6 +72,9 @@ public class SyncDataProviders: DataProvidersSource {
             .sink { [weak self] isSyncDisabled in
                 self?.credentialsAdapter.cleanUpDatabaseAndUpdateSchedule(shouldEnable: isSyncDisabled)
                 self?.bookmarksAdapter.cleanUpDatabaseAndUpdateSchedule(shouldEnable: isSyncDisabled)
+                if isSyncDisabled {
+                    self?.bookmarksAdapter.handleFavoritesAfterDisablingSync()
```
Comment (ayoy) [inline Core/SyncDataProviders.swift]: Consider moving this to bookmarksAdapter.cleanUpDatabaseAndUpdateSchedule

File: DuckDuckGo/AddOrEditBookmarkViewController.swift [hunk 3]
```diff
@@ -51,6 +51,7 @@ class AddOrEditBookmarkViewController: UIViewController {
         self.bookmarksDatabase = bookmarksDatabase
         self.viewModel = BookmarkEditorViewModel(editingEntityID: editingEntityID,
                                                  bookmarksDatabase: bookmarksDatabase,
+                                                 favoritesDisplayMode: AppDependencyProvider.shared.appSettings.favoritesDisplayMode,
```
Comment (ayoy) [inline DuckDuckGo/AddOrEditBookmarkViewController.swift]: Try parametrize appSettings here.

File: DuckDuckGo/AppDelegate.swift [hunk 4]
```diff
@@ -159,6 +159,16 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
                                                   to: context)
             legacyStorage?.removeStore()
 
+            do {
+                BookmarkUtils.migrateToFormFactorSpecificFavorites(byCopyingExistingTo: .mobile, in: context)
+                if context.hasChanges {
+                    try context.save()
+                }
+            } catch {
+                Thread.sleep(forTimeInterval: 1)
```
Comment (ayoy) [inline DuckDuckGo/AppDelegate.swift:200]: Send pixel before fatal erroring

File: DuckDuckGo/RemoteMessaging.swift [hunk 5]
```diff
@@ -97,24 +104,28 @@ struct RemoteMessaging {
...
 
     /// Convenience function
-    static func fetchAndProcess(bookmarksDatabase: CoreDataDatabase) async throws {
+    static func fetchAndProcess(bookmarksDatabase: CoreDataDatabase, favoritesDisplayMode: FavoritesDisplayMode) async throws {
         
         var bookmarksCount = 0
         var favoritesCount = 0
         let context = bookmarksDatabase.makeContext(concurrencyType: .privateQueueConcurrencyType)
         context.performAndWait {
+            let mobileFavoritesFolder = BookmarkUtils.fetchFavoritesFolder(withUUID: favoritesDisplayMode.displayedFolder.rawValue, in: context)!
```
Comment (ayoy) [inline DuckDuckGo/RemoteMessaging.swift]: s/mobile/displayed

File: DuckDuckGo/BookmarkFoldersTableViewController.swift [hunk 6]
```diff
@@ -159,7 +159,7 @@ class BookmarkFoldersViewController: UITableViewController {
             fatalError("Failed to dequeue \(FavoriteCell.reuseIdentifier) as FavoriteCell")
         }
 
-        cell.favoriteToggle.isOn = viewModel?.bookmark.isFavorite == true
+        cell.favoriteToggle.isOn = viewModel?.bookmark.isFavorite(on: .mobile) == true
```
Comment (ayoy) [inline DuckDuckGo/BookmarkFoldersTableViewController.swift]: fix it

File: DuckDuckGo/BookmarksDataSource.swift [hunk 7]
```diff
@@ -52,7 +52,7 @@ class BookmarksDataSource: NSObject, UITableViewDataSource {
             let cell = BookmarksViewControllerCellFactory.makeBookmarkCell(tableView, forIndexPath: indexPath)
             cell.faviconImageView.loadFavicon(forDomain: bookmark.urlObject?.host, usingCache: .fireproof)
             cell.titleLabel.text = bookmark.title
-            cell.favoriteImageViewContainer.isHidden = !bookmark.isFavorite
+            cell.favoriteImageViewContainer.isHidden = !bookmark.isFavorite(on: .mobile)
```
Comment (ayoy) [inline DuckDuckGo/BookmarksDataSource.swift]: fix it

File: DuckDuckGo/BookmarksViewController.swift [hunk 8]
```diff
@@ -113,14 +115,21 @@ class BookmarksViewController: UIViewController, UITableViewDelegate {
...
+          syncDataProviders: SyncDataProviders,
+          appSettings: AppSettings
     ) {
         self.bookmarksDatabase = bookmarksDatabase
         self.searchDataSource = SearchBookmarksDataSource(searchEngine: bookmarksSearch)
-        self.viewModel = BookmarkListViewModel(bookmarksDatabase: bookmarksDatabase, parentID: parentID, syncService: syncService)
+        self.viewModel = BookmarkListViewModel(
+            bookmarksDatabase: bookmarksDatabase,
+            parentID: parentID,
+            favoritesDisplayMode: AppDependencyProvider.shared.appSettings.favoritesDisplayMode,
```
Comment (ayoy) [inline DuckDuckGo/BookmarksViewController.swift]: replace with appsettings

File: DuckDuckGo/BookmarksViewController.swift [hunk 9]
```diff
@@ -286,8 +307,8 @@ class BookmarksViewController: UIViewController, UITableViewDelegate {
         let cell = tableView.cellForRow(at: indexPath)
         cell?.tintColor = .black
 
-        let title = bookmark.isFavorite ? UserText.actionRemoveFavorite : UserText.favorite
-        let iconName = bookmark.isFavorite ? "Favorite-Remove-24" : "Favorite-24"
+        let title = bookmark.isFavorite(on: .mobile) ? UserText.actionRemoveFavorite : UserText.favorite
```
Comment (ayoy) [inline DuckDuckGo/BookmarksViewController.swift]: fix

File: DuckDuckGo/FavoritesDisplayMode+UserDefaults.swift [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        case .displayNative:
+            return "display_native"
+        case .displayUnified:
+            return "display_all"
+        }
+    }
+
+    public init?(_ description: String) {
+        switch description {
+        case "display_native":
```
Comment (ayoy) [inline DuckDuckGo/FavoritesDisplayMode+UserDefaults.swift]: `case FavoritesDisplayMode.displayNative(.mobile).description:`

File: DuckDuckGo/FavoritesDisplayMode+UserDefaults.swift [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+//  limitations under the License.
+//
+
+import Bookmarks
+import Foundation
+
+extension FavoritesDisplayMode: LosslessStringConvertible {
+    static let `default` = FavoritesDisplayMode.displayNative(.mobile)
+
+    public var description: String {
```
Comment (ayoy) [inline DuckDuckGo/FavoritesDisplayMode+UserDefaults.swift]: Move this to BSK (CustomStringConvertible)

File: DuckDuckGo/FavoritesDisplayModeSyncHandler.swift [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+import Bookmarks
+import Combine
+import Foundation
+import SyncDataProviders
+
+extension SettingsProvider.Setting {
+    static let favoritesDisplayMode = SettingsProvider.Setting(key: "favorites_display_mode")
+}
+
+final class FavoritesDisplayModeSyncHandler: SettingSyncHandler {
```
Comment (ayoy) [inline DuckDuckGo/FavoritesDisplayModeSyncHandler.swift]: Make abstract common class for iOS and macOS implementation, defining the setting.

File: DuckDuckGo/MainViewController.swift [hunk 13]
```diff
@@ -340,6 +350,20 @@ class MainViewController: UIViewController {
...
     }
 
+    private func bindFavoritesDisplayMode() {
+        favoritesDisplayModeCancellable = NotificationCenter.default.publisher(for: AppUserDefaults.Notifications.favoritesDisplayModeChange)
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] _ in
+                guard let self else {
+                    return
+                }
+                self.menuBookmarksViewModel.favoritesDisplayMode = self.appSettings.favoritesDisplayMode
```
Comment (ayoy) [inline DuckDuckGo/MainViewController.swift:506]: Check if this is needed, possibly update menu bookmarks view model to reload data.

File: DuckDuckGo/RemoteMessaging.swift [hunk 14]
```diff
@@ -97,24 +104,28 @@ struct RemoteMessaging {
...
+
             let bookmarksCountRequest = BookmarkEntity.fetchRequest()
-            bookmarksCountRequest.predicate = NSPredicate(format: "%K == nil AND %K == false AND %K == false",
-                                                 #keyPath(BookmarkEntity.favoriteFolder),
-                                                 #keyPath(BookmarkEntity.isFolder),
-                                                 #keyPath(BookmarkEntity.isPendingDeletion))
+            bookmarksCountRequest.predicate = NSPredicate(format: "!(ANY %K CONTAINS %@) AND %K == false AND %K == false",
```
Comment (ayoy) [inline DuckDuckGo/RemoteMessaging.swift]: should be `!(%K CONTAINS %@)`

File: DuckDuckGo/RemoteMessaging.swift [hunk 15]
```diff
@@ -97,24 +104,28 @@ struct RemoteMessaging {
...
+                                                          #keyPath(BookmarkEntity.isFolder),
+                                                          #keyPath(BookmarkEntity.isPendingDeletion))
             bookmarksCount = (try? context.count(for: bookmarksCountRequest)) ?? 0
             
             let favoritesCountRequest = BookmarkEntity.fetchRequest()
-            bookmarksCountRequest.predicate = NSPredicate(format: "%K != nil AND %K == false AND %K == false",
-                                                 #keyPath(BookmarkEntity.favoriteFolder),
-                                                 #keyPath(BookmarkEntity.isFolder),
-                                                 #keyPath(BookmarkEntity.isPendingDeletion))
+            favoritesCountRequest.predicate = NSPredicate(format: "ANY %K CONTAINS %@ AND %K == false AND %K == false",
```
Comment (ayoy) [inline DuckDuckGo/RemoteMessaging.swift]: should be `%K CONTAINS %@)`

File: DuckDuckGo/RemoteMessaging.swift (hunk 14 above)
Comment (ayoy) [reply DuckDuckGo/RemoteMessaging.swift]: Turns out it needs to be `SUBQUERY(%K, $x, $x CONTAINS %@).@count == 0` (just for the negated predicate, the favorites predicate `%K CONTAINS %@` works fine)

File: DuckDuckGo/MainViewController.swift (hunk 13 above)
Comment (ayoy) [reply DuckDuckGo/MainViewController.swift:506]: This call actually invalidates favoritesRoot in MenuBookmarksViewModel, so we can keep it here.

**Comment by ayoy** [comment]
Adding a "Pending Product Review" label (in absence of "DO NOT MERGE") so that it can't be accidentally merged.

**Review by SabrinaTardio — APPROVED** [review]

**Pull request merged**